### DISCUSSION TITLE
Any success with MicroPython on the LilyGo T-Internet-POE?

### DISCUSSION THREAD

**Discussion by bjpirt** [micropython/micropython#12314] [ESP32]
I've been trying to get MicroPython working on one of these boards: https://www.lilygo.cc/products/t-internet-poe but with no success. The board is working with Arduino so I know the hardware is good, however when I flash MicroPython onto it and try to initialise the LAN I can't seem to find a combination of parameters that works.

What I know:
 - MDC Pin = 23
 - MDIO Pin = 18
 - The clock is on pin 17 output (like the Olimex boards)
 - From what I can tell, there is now power enable pin
 
I've tried the following:

```
import network, machine
l = network.LAN(mdc=machine.Pin(23), mdio=machine.Pin(18), power=None, phy_type=network.PHY_LAN8720, phy_addr=0, ref_clk=machine.Pin(17), ref_clk_mode=machine.Pin.OUT)
```
But this just gives me:
```
TypeError: extra keyword arguments given
```

If I try it without the `ref_clk` arguments (which are what seem to be causing this error), I don't get a TypeError, but instead get:

```
E (8560) emac_esp32: emac_esp32_init(307): reset timeout
E (8560) esp_eth: esp_eth_driver_install(198): init mac failed
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: esp_eth_driver_install failed
```

Does anyone have any pointers on how I might be able to get this working please?

**Comment by robert-hh** [comment]
Which firmware do you use?

**Comment by robert-hh** [comment]
I found schematics for a LiLiGo POE, where there is a crystal at the LAN8720 for refclk, and thus refclk at the ESP32 should be an input. But in any case it mus be possible to define these signals. You could try to use the Olimex ESP32 POE firmware.

**Comment by bjpirt** [comment]
@robert-hh I'm using the generic esp32 1.20 firmware.

I'll give the olimex PoE firmware a shot

**Reply by robert-hh** [reply]
Thinking about the LiLi-T Circuit, even if there is a crystal, it cannot drive the ESP32 input. So refclk_mode should probably be OUT.

**Comment by bjpirt** [comment]
@robert-hh I've tried the olimex build which gives exactly the same results unfortunately - compalins about extra keyword args.

It's odd that this is the error that's getting thrown as I've looked through the source and all of these args look like they should be fine. It's not even getting to trying to initialise the lan.

**Comment by bjpirt** [comment]
I just tried stripping out all of the arguments for the call to lan() and adding them back in as it complained and I've found that the following works:

```
l = network.LAN(mdc=machine.Pin(23), mdio=machine.Pin(18), phy_type=network.PHY_LAN8720, phy_addr=0)
```

Not sure why the other arguments caused it to fail, but at least this now seems to be working (on the olimex build - will try on the default now)

**Comment by bjpirt** [comment]
Just tried it on the default build and it doesn't work. Not sure what the difference between the two is but I think it _should_ work on the default build so I'll see if I can fix that

**Comment by robert-hh** [comment]
It is strange that the `ref_clk` and `ref_clk_mode` arguments are rejected. Besides that, the default settings for ref_clk_mode is IN, and the ref_clk gpio is not set. What is the version string that you get, when pushing Ctrl-B in REPL?

Edit: The one I see with the lastest nightly build is:
MicroPython v1.20.0-92-g3a7ad6426-dirty on 2023-05-19; Olimex ESP32 ETH with ESP32

**Comment by robert-hh** [comment]
V1.20.0 rejects the `ref_clk` and `ref_clk_mode` keywords. That's strange, since it was added before the change to v1.20.0, and it worked before. Nevertheless, the nightly build works.
Edit: the GENERIC version of the nightly build works as well with the Olimex board.

**Comment by robert-hh** [comment]
> That's strange, since it was added before the change to v1.20.0

Not strange. Setting ref_clk etc. requires esp-idf 4.4 and newer. V1.20.0 was built with esp-idf v4.2

**Comment by bjpirt** [comment]
Ah, that makes sense - will use the nightly build. Thanks for your help with this

**Comment by bjpirt** [comment]
Hmm, when I try the nightly build I get the following:

```
>>> import network, machine
>>> l = network.LAN(mdc=machine.Pin(23), mdio=machine.Pin(18), phy_type=network.PHY_LAN8720, phy_addr=0)
E (67680) eth_phy_802_3: eth_phy_802_3_pwrctl(162): power up timeout
E (67680) eth_phy_802_3: esp_eth_phy_802_3_basic_phy_init(335): power control failed
E (67680) lan87xx: lan87xx_init(292): failed to init PHY
E (67690) esp_eth: esp_eth_driver_install(229): init phy failed
```
I also tried with `power=None` but had the same result. Not sure if I should set power to something else as afaict there is no power control on the board

**Comment by bjpirt** [comment]
OK, got it working with the nightly now too, it just needed the `ref_clk` arguments as well, unlike the Olimex build. Here's what I've got working now for others' reference:

```
>>> import machine, network
>>> l = network.LAN(mdc=machine.Pin(23), mdio=machine.Pin(18), power=None, phy_type=network.PHY_LAN8720, phy_addr=0, ref_clk=machine.Pin(17), ref_clk_mode=machine.Pin.OUT)
>>> l.active(True)
>>> l.ifconfig()
('192.168.88.254', '255.255.255.0', '192.168.88.1', '192.168.88.1')
```
